### PR TITLE
deterministic tie-breaking between categories.

### PR DESCRIPTION
**Pull request by glycerine**
CatBallotBox and its Tally() is here modified to use a slice rather than a map.
   This avoids Go's non-deterministic map range, which made tie-breaking
   between categories somewhat random, with the result that repeated
   evaluations of the same tree could produce different predictions and
   thus different error rates.

   Test wise, lytics' fork has TestIris in data_test.go
   using tree Copy()-ing that will sporadically fail without this fix.

### CODE REVIEW

**Comment by glycerine** [comment]
same as https://github.com/ryanbressler/CloudForest/pull/64, but rebased against the Lytics' fork for convenience.

I think this a cleaner, faster approach, when compared to https://github.com/lytics/CloudForest/pull/18

File: catballotbox.go [hunk 1]
```diff
@@ -39,12 +53,13 @@ func NewCatBallotBox(size int) *CatBallotBox {
 //Vote registers a vote that case "casei" should be predicted to be the
 //category "pred".
 func (bb *CatBallotBox) Vote(casei int, pred string, weight float64) {
+
 	predn := bb.CatToNum(pred)
 	bb.Box[casei].Mutex.Lock()
-	if _, ok := bb.Box[casei].Map[predn]; !ok {
-		bb.Box[casei].Map[predn] = 0
+	for len(bb.Box[casei].Cat) <= predn {
```
Comment (timkaye11) [inline catballotbox.go:59]: I think `predn` is always sequential, so could the `for` be replaced with an `if`?

**Comment by timkaye11** [comment]
thanks for working on this @glycerine!

File: catballotbox.go (hunk 1 above)
Comment (glycerine) [reply catballotbox.go:59]: it panics with an `if` instead of a `for`, panic: runtime error: index out of range, on TestThreeTargets, for example.

**Comment by timkaye11** [comment]
thanks again for working on this @glycerine! After reading through this a little bit, I agree that this PR is a cleaner and faster approach compared to  #18, but it could also introduce more memory usage. 

Consider a dataset where there are 100+ different classes in a classification model. If the `predn` below was the 100th class, then we'd have to zero-pad the `Cat` array 100 times, whereas a map would only need to store a single value. 

https://github.com/lytics/CloudForest/blob/356fbf5bd31f19e506c545f3c6528a0c2b25165d/catballotbox.go#L59-L61

Basically it comes down to a performance/memory tradeoff. At Lytics, lots of our classification models tend to have many classes, so IMO sacrificing a slight performance downgrade (due to `sort.Slice`) is preferable to increased memory usage

**Comment by glycerine** [comment]
Thanks for your review and work on the issue, @timkaye11 

>map would only need to store a single value.

Doesn't this assume that none of the other other categories ever get a vote? Seems very unlikely. Typically most if not all categories end up getting at least one vote from some tree. 

Anyway, this is splitting hairs and I think that keeping the map version is less risky in terms of correctness too.  I'll close this.

**Comment by glycerine** [comment]
given the logic issues in the #18, I'll reopen for consideration and correctness comparison.

**Comment by glycerine** [comment]
I stress tested compared and these two seem. Equivalent. The logic bug wasn't real after all. Closing this out now that #18 is merged.
